dropshot/src/server.rs

```diff
@@ -173,6 +152,85 @@ impl HttpServer {
     }
 }
 
+/**
+ * A running Dropshot HTTP server.
+ */
+pub struct HttpServer {
+    app_state: Arc<DropshotState>,
+    local_addr: SocketAddr,
+    join_handle: Option<tokio::task::JoinHandle<Result<(), hyper::Error>>>,
+    close_channel: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl HttpServer {
+    pub fn local_addr(&self) -> SocketAddr {
+        self.local_addr
+    }
+
+    pub fn app_private(&self) -> Arc<dyn Any + Send + Sync + 'static> {
+        Arc::clone(&self.app_state.private)
+    }
+
+    /**
+     * Signals the currently running server to stop (if it hasn't
+     * stopped already) and waits for it to exit.
+     */
+    pub async fn close(mut self) -> Result<(), String> {
+        self.close_channel
+            .take()
+            .expect("cannot close twice")
+            .send(())
+            .expect("failed to send close signal");
+        if let Some(handle) = self.join_handle.take() {
+            let join_result = handle
+                .await
+                .map_err(|error| format!("waiting for server: {}", error))?;
+            join_result.map_err(|error| format!("server stopped: {}", error))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/**
+ * Dropping the HttpServer will terminate the server, without actually waiting
+ * for it to stop.
+ *
+ * For graceful termination, use the `close` function instead.
+ */
+impl Drop for HttpServer {
```

Comment:
What would happen if we didn't impl `Drop` here?

Reply:
If a client let an `HttpServer` object go out of scope without calling `await` *and* without calling `close`, it would be impossible to close the server. It would be running forever, and there would be no way to tell it to shut down.

I'm open to discuss this - personally, I felt "advise the server to terminate, even though we can't await when it does shut down" is better than "leak it forever", but I suppose we could delete this `Drop` impl and leak-by-default. That being said, I acknowledge this "leak when you let things go out of scope" behavior is seen elsewhere, such as with Tokio's JoinHandles.

Reply:
That's interesting.  I believe that before this change, if the server got dropped, it was shutdown.  (I ran into a bug with this a few months ago.)

I don't have a strong feeling about desired behavior here, but I also don't feel I fully grok the implications of _not_ doing this (e.g., for things like cancellation or panic unwinding).

Is it possible to log about this case?  If it's a reasonable thing to do, maybe just at "info".  If not, maybe "warn" -- or maybe even panic?

Reply:
I'm not sure I see how the server would be shutdown before this change - the only connection to the server is via the tokio::task joinhandle, which explicitly does *not* stop the background task when dropped.

Regardless, added a "warn" statement to the drop implementation.

Reply:
> I'm not sure I see how the server would be shutdown before this change - the only connection to the server is via the tokio::task joinhandle, which explicitly does not stop the background task when dropped.

That's true, but the guts of the Dropshot server is really in the Hyper server that it owns, and that may shut itself down on `drop`.

To give a little more detail about the problem I saw before: it related to this kind of test in oxide-api-prototype:
https://github.com/oxidecomputer/oxide-api-prototype/blob/d254e720b20cae2c8076ac196023a51af0f1fb6c/tests/test_instances.rs#L39

That `test_setup` function is this one:
https://github.com/oxidecomputer/oxide-api-prototype/blob/d254e720b20cae2c8076ac196023a51af0f1fb6c/tests/common/mod.rs#L46-L95

It returns a struct that owns the server and (prior to this change) the join handle.

At one point, the line I linked above (test_instances.rs L39) was instead something like `let client = test_setup("test_instances").await.external_client` (i.e., combining lines 39 and 40).  I found that when I did that, tests failed because the server was shutdown immediately, right after that call.  @cbiffle and I debugged this to see that `drop` was being invoked on the server because we weren't hanging onto the `ControlPlaneTestContext` struct or referencing the server later.  It's possible something else changed in the meantime, as I can't reproduce this by making that change to oxide-api-prototype#master.

Do you think it's worth testing the behavior now without this impl to see if we can avoid it?

Reply:
Sure, will test and report back before merging

Reply:
So I tried making a couple tests:

Baseline, to make sure "the normal case works":
1. Create an `ApiDescription` with a simple callback.
2. Invoke `HttpServerStarter::new(...).unwrap().start()`.
3. Keep the server object alive
4. Make a client request --> Observe the result is OK

The drop-server case:
1. Create an `ApiDescription` with a simple callback.
2. Invoke `HttpServerStarter::new(...).unwrap().start()`.
3. Drop the server
4. Make a client request --> Expect it to fail  (the server should be terminated).

*without* the `impl Drop`, the `tokio::spawn`-ed task panics, as it reports "dropshot server shutting down without invoking close()". The client request then fails, as the server thread has panicked.

*with* the `impl Drop`, the `tokio::spawn`-ed task no longer panics - instead, it emits the warning I added ("dropped HttpServer without calling close or await"), and the client request fails.

TL;DR: I *think* the trigger for the server exiting on termination is actually the `tokio::sync::oneshot::Sender` object going out-of-scope, which causes the "graceful shutdown" to trigger with an error.

Regardless, it seems this behavior is covered, and if we consider it an error, I'll happily remove the `impl Drop` and add documentation to the `HttpServer` struct instead, clarifying that it must be `close`-ed or `await`-ed to completion (or else it'll panic).

Reply:
One extra layer of complexity: if the `tokio::spawn`-ed task panics as a result of this "server-out-of-scope" issue, it actually [won't stop execution](https://github.com/tokio-rs/tokio/issues/700) if the default tokio runtime is used. I noticed this "panic in the spawned task" behavior actually wouldn't cause execution to stop / tests to fail.

To me, this is a *big* red flag - if a task can panic, but execution remains uninterrupted, it would be easy to misuse the API without receiving any meaningful notification. I've updated the implementation of drop to flag this case more explicitly to the task holding the server, and added some test cases to validate this behavior.

Reply:
> TL;DR: I think the trigger for the server exiting on termination is actually the tokio::sync::oneshot::Sender object going out-of-scope, which causes the "graceful shutdown" to trigger with an error.

That rings a bell -- yes, I think this was how things shut down in the test case I was describing earlier.

> Regardless, it seems this behavior is covered, and if we consider it an error, I'll happily remove the impl Drop and add documentation to the HttpServer struct instead, clarifying that it must be close-ed or await-ed to completion (or else it'll panic).

That makes the most sense to me (and it's why I added that assertion in the first place).  I also have the feeling that folks coming from C++ and maybe mainstream Rust will find it surprising.

> One extra layer of complexity: if the tokio::spawn-ed task panics as a result of this "server-out-of-scope" issue, it actually won't stop execution if the default tokio runtime is used. I noticed this "panic in the spawned task" behavior actually wouldn't cause execution to stop / tests to fail...To me, this is a big red flag - if a task can panic, but execution remains uninterrupted, it would be easy to misuse the API without receiving any meaningful notification.

I think this is an instance of a more general problem.  My strong preference would be that we treat all Rust panics as programmer errors and build all of our own software with panic=abort so that they abort(3c) and dump core on panic.

Reply:
I agree with you on the goal to "treat all Rust panics as programmer errors". An `impl Drop` still exists, but only to force the main task to panic if close has not been invoked (it does not attempt to trigger a close if it hasn't happened already). I consider this acceptable, because the alternative is a panic in the background task, which may or may not be visible depending on panic=abort behavior.

Marking resolved, happy to re-open if that seems controversial.

Reply:
I'm having trouble finding documentation about tokio's behavior here, but doesn't this behavior rely on the caller running the server itself in the "main" task?